A60-xds-stateful-session-affinity-weighted-clusters.md

```diff
@@ -0,0 +1,270 @@
+     or weighted-clusters and cluster should match/be present in one of those.
+
+ * If validation succeeds, then override the cluster call-attribute (originally
+   set by the config selector) with the cluster name.
+
+ * Set the override-host value with the IP:port value from the cookie (this is
+   the current logic).
+
+### New `ClusterSelectionFilter`
+
+Because of the changes we made to the config selector we need a new filter
+to finalize cluster selection and to release refs to the cluster map and
+route config. This new filter is called `ClusterSelectionFilter` and it runs
+just before the Router filter. This is an internal implementation detail of
+gRPC so the xDS config is not aware of this new filter. Note:
+
+ * This is the second-last filter (retry/router filter being last).
+   Specifically it has to run after the `StatefulSessionFilter`.
+
+ * It gets the selected cluster name from `XdsClusterAttributeTypeName` and
+   cluster map from `XdsClusterMapAttribute`.
+
+ * It then takes a ref to that cluster entry from the map and releases the
+   ref to the cluster map.
+
+With this approach, `StatefulSessionFilter` is able to update the selected
+cluster between when the config selector selects the cluster and when the
+`ClusterSelectionFilter` takes a ref to the cluster.
+
+### Process Diagram
+
+The following diagram shows how the process works once this design is
+implemented:
+
+![Weighted Clusters Process](A60_graphics/final-diagram.png)
+
+ 1. An RPC comes in and config selector picks cluster-A (shown in bold).
+
+ 2. config selector sets 3 attributes to expose the cluster, current route,
+    cluster map and the entire route config. Note it also takes ref to the
+    cluster map and the route config.
+
+ 3. HTTP filters are invoked - specifically the stateful session affinity
+    filter that is of interest to us.
+
+ 4. the stateful session affinity filter sets the override-host attribute
+    and it also overwrites the cluster-attr based on the cookie contents.
+
+ 5. the new cluster selection filter is invoked.
+
+ 6. the cluster selection filter takes a ref to the new cluster - cluster-B
+    (shown in bold and italics) and releases the ref to the whole map and
+    the route config.
+
+ 7. Now the cluster-manager LB is invoked. It reads the cluster attribute -
+    which happens to be the correct cluster.
+
+ 8. After the usual LB hierarchy processing, the override-host LB policy is
+    invoked.
+
+ 9. The policy reads the override-host attribute and uses it to route the
+    RPC to that host which is now in the correct cluster.
+
+### Notes
+
+This design mainly deals with how two existing features - stateful session
+affinity and weighted clusters - should work together. As a result this feature
+does not require a new API or impact any existing APIs.
+
+### Temporary environment variable protection
+
+[A55: xDS-Based Stateful Session Affinity for Proxyless gRPC][A55] was only
```

Comment:
I think the users (like @va7ish and others) mainly want to use stateful session affinity (SSA) with weighted clusters support and without that support they probably consider the feature as incomplete. Since the basic SSA feature is already there, users may do limited testing but will wait for  weighted clusters support before using it or doing full testing.

In any case let's wait for input from @va7ish .

Reply:
@va7ish could you provide your input here?

Reply:
Yes, I think we should consider SSA incomplete if the weighted traffic routing configuration breaks SSA.

Reply:
So just to confirm and to answer @markdroth 's question in the first comment: it is unlikely that you will use the existing stateful session affinity support without the new weighted cluster logic that is being added. So a single env var to guard both sets of code should be sufficient according to you?

Reply:
Okay, super.  Then using the same env var guard sounds good.